pkg/internal/testing/certs/tinyca_test.go

```diff
@@ -0,0 +1,254 @@
+		firstCerts, err := ca.NewServingCert()
+		Expect(err).NotTo(HaveOccurred())
+		secondCerts, err := ca.NewClientCert(certs.ClientInfo{Name: "user"})
+		Expect(err).NotTo(HaveOccurred())
+		thirdCerts, err := ca.NewServingCert()
+		Expect(err).NotTo(HaveOccurred())
+
+		By("checking that they have different serials")
+		serials := []*big.Int{
+			firstCerts.Cert.SerialNumber,
+			secondCerts.Cert.SerialNumber,
+			thirdCerts.Cert.SerialNumber,
+		}
+		// quick uniqueness check of numbers: sort, then you only have to compare sequential entries
+		sort.Slice(serials, func(i, j int) bool {
+			return serials[i].Cmp(serials[j]) == -1
+		})
+		Expect(serials[1].Cmp(serials[0])).NotTo(Equal(0), "serials shouldn't be equal")
+		Expect(serials[2].Cmp(serials[1])).NotTo(Equal(0), "serials shouldn't be equal")
+	})
+
+	Describe("Generated serving certs", func() {
+		It("should be valid for short enough to avoid production usage, but long enough for long-running tests", func() {
+			cert, err := ca.NewServingCert()
+			Expect(err).NotTo(HaveOccurred(), "should be able to generate the serving certs")
+
+			duration := cert.Cert.NotAfter.Sub(time.Now())
+			Expect(duration).To(BeNumerically("<=", 168*time.Hour), "not-after should be short-ish (<= 1 week)")
+			Expect(duration).To(BeNumerically(">=", 2*time.Hour), "not-after should be enough for long tests (couple of hours)")
+		})
+
+		Context("when encoding names", func() {
+			var cert certs.CertPair
+			BeforeEach(func() {
+				By("generating a serving cert with IPv4 & IPv6 addresses, and DNS names")
+				var err error
+				// IPs are in the "example & docs" blocks for IPv4 (TEST-NET-1) & IPv6
+				cert, err = ca.NewServingCert("192.0.2.1", "localhost", "2001:db8::")
+				Expect(err).NotTo(HaveOccurred(), "should be able to create the serving certs")
+			})
+
+			It("should encode all non-IP names as DNS SANs", func() {
+				Expect(cert.Cert.DNSNames).To(ConsistOf("localhost"))
+			})
+
+			It("should encode all IP names as IP SANs", func() {
+				// NB(directxman12): this is non-exhaustive because we also
+				// convert DNS SANs to IPs too (see test below)
+				Expect(cert.Cert.IPAddresses).To(ContainElements(
+					// normalize the elements with To16 so we can compare them to the output of
+					// of ParseIP safely (the alternative is a custom matcher that calls Equal,
+					// but this is easier)
+					WithTransform(net.IP.To16, Equal(net.ParseIP("192.0.2.1"))),
+					WithTransform(net.IP.To16, Equal(net.ParseIP("2001:db8::"))),
+				))
+			})
+
+			It("should add the corresponding IP address(es) (as IP SANs) for DNS names", func() {
+				// NB(directxman12): we currently fail if the lookup fails.
+				// I'm not certain this is the best idea (both the bailing on
+				// error and the actual idea), so if this causes issues, you
+				// might want to reconsider.
+
+				localhostAddrs, err := net.LookupHost("localhost")
+				Expect(err).NotTo(HaveOccurred(), "should be able to find IPs for localhost")
```

Comment:
Are you sure you want to do this lookup? I have seen it returning very surprising results at times :P

Reply:
we actually do this currently within the code (perhaps unfortunately, depending on how you're looking at it), so it's there for compat reasons.  This just adds a test for that behavior.